src/solvers/flattening/boolbv_equality.cpp

```diff
@@ -21,9 +21,11 @@ literalt boolbvt::convert_equality(const equal_exprt &expr)
 {
   if(!base_type_eq(expr.lhs().type(), expr.rhs().type(), ns))
   {
-    std::cout << "######### lhs: " << expr.lhs().pretty() << '\n';
-    std::cout << "######### rhs: " << expr.rhs().pretty() << '\n';
-    throw "equality without matching types";
+    error().source_location=expr.find_source_location();
+    error() << "equality without matching types; \n"
```

Comment:
1. Replace `;` by `:`? No space before newline, please. (If anyone is pattern matching on those error messages they would easily get confused when trying to match on the full line.)
2. This entire check should be moved to `goto-symex` and should just be a `DATA_INVARIANT` using a plain equality test (no `base_type_eq`) here. This error does show up in some situations when pointer dereferencing messed up, but diagnosing it here is way too late as it provides poor customer experience and makes debugging impossible (the way to debug it is placing a check in goto-symex - been there, done that).